lib/autoupdate/rolloutcontroller/reconciler.go

```diff
@@ -0,0 +1,222 @@
+	"sync"
+	"time"
+)
+
+const (
+	reconciliationTimeout = 30 * time.Second
+	defaultConfigMode     = update.AgentsUpdateModeEnabled
+	defaultStrategy       = update.AgentsStrategyHaltOnError
+)
+
+// Reconciler reconciles the AutoUpdateAgentRollout singleton based on the content of the AutoUpdateVersion and
+// AutoUpdateConfig singletons. This reconciler is not based on the services.GenericReconciler because:
+// - we reconcile 2 resources with one
+// - both input and output are singletons, we don't need the multi resource logic nor stream/paginated APIs
+type Reconciler struct {
+	clt Client
+	log *slog.Logger
+
+	// lock ensures we only run one reconciliation at a time
+	lock sync.Mutex
+}
+
+// Reconcile the AutoUpdateAgentRollout singleton. The reconciliation can fail because of a conflict (multiple auths
+// are racing), in this case we retry the reconciliation immediately.
+func (r Reconciler) Reconcile(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, reconciliationTimeout)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			err := r.tryReconcile(ctx)
+			switch {
+			case err == nil:
+				return nil
+			case trace.IsCompareFailed(err), trace.IsNotFound(err):
+				// The resource changed since we last saw it
+				// We must have raced against another auth
+				// Let's retry the reconciliation
+				// TODO: add a backoff to avoid a tight or infinite loop?
+				r.log.DebugContext(ctx, "retrying reconciliation", "error", err)
+			default:
+				// error is non-nil and non-retryable
+				return trace.Wrap(err, "failed to reconcile rollout")
+			}
+		}
+	}
+}
+
+// tryReconcile tries to reconcile the AutoUpdateAgentRollout singleton.
+// This function should be nilpotent if the AutoUpdateAgentRollout is already up-to-date.
+// The creation/update/deletion can fail with a trace.CompareFailedError or trace.NotFoundError
+// if the resource change while we were computing it.
+// The caller must handle those error and retry the reconciliation.
+func (r Reconciler) tryReconcile(ctx context.Context) error {
+	// get autoupdate_config
+	config, err := r.clt.GetAutoUpdateConfig(ctx)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err, "getting autoupdate_config")
+	}
+
+	// get autoupdate_version
+	version, err := r.clt.GetAutoUpdateVersion(ctx)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err, "getting autoupdate_version")
+	}
+
+	// get autoupdate_agent_rollout
+	rolloutExists := true
+	existingRollout, err := r.clt.GetAutoUpdateAgentRollout(ctx)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err, "getting autoupdate_agent_rollout")
+	}
+	if trace.IsNotFound(err) {
+		// rollout doesn't exist yet, we'll need to call Create instead of Update.
+		rolloutExists = false
+	}
+
+	// if autoupdate_version does not exist or does not contain spec.agents, we should not configure a rollout
+	if version.GetSpec().GetAgents() == nil {
```

Comment:
Technically `version` and `config` shouldn't be touched (except for resetting) if the get functions returned an error, so the `rolloutExists` thing should also be extended to `version` and `config`.